### DISCUSSION TITLE
Compress PDF file

### DISCUSSION THREAD

**Discussion by Mano2731995** [pymupdf/PyMuPDF#2107] [Looking for help] [answered]
Hai jorj

For example I am having more than 500MB pdf file. How to compress that file in fitz?

**Comment by JorjMcKie** [comment]
The standard approach is using garbage collection and deflation.
You can do this with MuPDF's CLI utility `mutool clean -ggggz input.pdf output.pdf`.
Option z means deflation, the multiple "g" option corresponds to maximum garbage collection, argument `garbage=4` in PyMuPDF - fully specified `doc.save("output.pdf", garbage=4, deflate=True)`.

This will not change the PDF's appearance in any way.

Depending on your willingness to bargain appearance for size, more things can be done with PyMuPDF - and some might have a very significant effect:
* removal of embedded files
* subsetting of fonts if applicable
* replacing images by a smaller (and maybe lower quality) version, including color to grayscale, PNG to JPG, ...
* removal of images
* removal of page thumbnails if applicable
* ...

If any of the above finds your interest, we should discuss.

**Reply by Mano2731995** [reply]
Thanks for valuable reply jorj if required above query I will let u know. 

I need one help jorj I have attached file by using page.addfileannot.Fyr I have attached image. In that attachment having small size yellow comment symbol(red color circle) I need to ignore that yellow color comment symbol![Uploading d{e790b09} (6f61d9a4-a215-42ae-9e2f-3e227bdb0d24 id=0x7f0a03f6 tag=BaseCommentFragment) …]()

**Reply by JorjMcKie** [reply]
The attached link does not work.

**Reply by Mano2731995** [reply]
![IMG_20221208_141316.jpg](https://user-images.githubusercontent.com/117537492/206426564-5dbde432-a4ca-46c2-b672-4adbc191b8ad.jpg)

**Comment by JorjMcKie** [comment]
Hard to be sure what that is, because every viewer does these things differently.
But it **_looks_** like a PopUp annotation. They automatically get created if a "normal" annotation is clicked.
There exists method `Annot.delete_responses()` which removes PopUp annotations and other responses to the annotation.

**Comment by Mano2731995** [comment, marked as answer]
> Hard to be sure what that is, because every viewer does these things differently.
> But it **_looks_** like a PopUp annotation. They automatically get created if a "normal" annotation is clicked.
> There exists method `Annot.delete_responses()` which removes PopUp annotations and other responses to the annotation.

Superb jorj it's working as expected thank you

**Comment by JorjMcKie** [comment]
Yes, you need the image xref. You also must provide the new image of course. If you want a JPG version of the existing image, you must make it: there is no automatism for this in PyMuPDF - and in fact, why should there be such a thing?

You can extract the current image, make a pixmap from it, then use `pix.tobytes("jpg")` and reinsert this output to replace the current image.
________________________________
Von: kenho211 ***@***.***>
Gesendet: Donnerstag, 21. Dezember 2023 07:06
An: pymupdf/PyMuPDF ***@***.***>
Cc: Jorj X. McKie ***@***.***>; Comment ***@***.***>
Betreff: Re: [pymupdf/PyMuPDF] Compress PDF file (Discussion #2107)


How to replace the images from PNG to JPG?
Couldn't find relevant functions, do we have to do it using xref?
